utils/src/init_once.rs

```diff
@@ -0,0 +1,76 @@
+use core::cell::UnsafeCell;
+use core::sync::atomic::{self, AtomicBool};
+
+pub struct InitOnce<T: Sized> {
+    locked: AtomicBool,
+    val: UnsafeCell<Option<T>>,
+}
+
+impl<T: Sync> InitOnce<T> {
+    pub const fn new() -> Self {
+        Self {
+            locked: AtomicBool::new(false),
+            val: UnsafeCell::new(None),
+        }
+    }
+
+    pub fn set(&self, val: T) -> Result<(), &'static str> {
+        if self.is_initialized() {
+            return Err("the InitOnce is already initialized");
+        }
+
+        while self
+            .locked
+            .compare_exchange(
+                false,
+                true,
+                atomic::Ordering::Acquire,
+                atomic::Ordering::Relaxed,
+            )
+            .is_err()
+        {}
```

Comment:
I like the idea. Should we create a SpinLock type?

Reply:
In the close future very much yes, but I'd rather keep this as is for now

Reply:
Fine by me